opnode/l2/ssz.go

```diff
@@ -0,0 +1,225 @@
+package l2
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"sync"
+)
+
+// ExecutionPayload is the only SSZ type we have to marshal/unmarshal,
+// so instead of importing a SSZ lib we implement the bare minimum.
+// This is more efficient than RLP, and matches the L1 consensus-layer encoding of ExecutionPayload.
+
+// All fields (4s are offsets to dynamic data)
+const executionPayloadFixedPart = 32 + 20 + 32 + 32 + 256 + 32 + 8 + 8 + 8 + 8 + 4 + 32 + 32 + 4
+
+// MAX_TRANSACTIONS_PER_PAYLOAD in consensus spec
+const maxTransactionsPerPayload = 1 << 20
+
+// The payloads are small enough to read and write at once.
+// But this happens often enough that we want to avoid re-allocating buffers for this.
+var payloadBufPool = sync.Pool{New: func() any {
+	x := make([]byte, 0, 100_000)
+	return &x
+}}
+
+func (payload *ExecutionPayload) SizeSSZ() (full uint32) {
+	full = executionPayloadFixedPart + uint32(len(payload.ExtraData))
+	// One offset to each transaction
+	full += uint32(len(payload.Transactions)) * 4
+	// Each transaction
+	for _, tx := range payload.Transactions {
+		full += uint32(len(tx))
+	}
+	return full
+}
+
+// marshalBytes32LE returns the value of z as a 32-byte little-endian array.
+func marshalBytes32LE(out []byte, z *Uint256Quantity) {
+	_ = out[31] // bounds check hint to compiler
+	binary.LittleEndian.PutUint64(out[0:8], z[0])
+	binary.LittleEndian.PutUint64(out[8:16], z[1])
+	binary.LittleEndian.PutUint64(out[16:24], z[2])
+	binary.LittleEndian.PutUint64(out[24:32], z[3])
+}
+
+func unmarshalBytes32LE(in []byte, z *Uint256Quantity) {
+	_ = in[31] // bounds check hint to compiler
+	z[0] = binary.LittleEndian.Uint64(in[0:8])
+	z[1] = binary.LittleEndian.Uint64(in[8:16])
+	z[2] = binary.LittleEndian.Uint64(in[16:24])
+	z[3] = binary.LittleEndian.Uint64(in[24:32])
+}
+
+// MarshalSSZ encodes the ExecutionPayload as SSZ type
+func (payload *ExecutionPayload) MarshalSSZ(w io.Writer) (n int, err error) {
+	scope := payload.SizeSSZ()
+
+	buf := *payloadBufPool.Get().(*[]byte)
+	if uint32(cap(buf)) < scope {
+		buf = make([]byte, scope)
+	} else {
+		buf = buf[:scope]
+	}
+	defer payloadBufPool.Put(&buf)
+
+	offset := uint32(0)
+	copy(buf[offset:offset+32], payload.ParentHash[:])
+	offset += 32
+	copy(buf[offset:offset+32], payload.FeeRecipient[:])
+	offset += 20
```

Comment:
Did you switch the increase of the offset with where to copy the bytes to?

Reply:
Not sure if I understand your question here

Reply:
The line `copy(buf[offset:offset+32], payload.FeeRecipient[:])` is correct because `copy` copies the min of `len(src), len(dst)`. You're copying 20 bytes into a 32 byte buffer.